### ISSUE TITLE
SingleChildScrollView very slow  if has MultiTapGestureRecognizer in it when upgrade to flutter 1.9

### ISSUE THREAD

**Issue by bigzhu** [flutter/flutter#40480]
the struct like this:
```
SingleChildScrollView
    RichText
        TextSpan with MultiTapGestureRecognizer
```
flutter 1.7.8+hotfix.2 everything is ok.

when upgraded to flutter v1.9.1+hotfix.2 debug mode work fine, but release mode SingleChildScrollView very very slow almost can't scroll.

if remove MultiTapGestureRecognizer, it scrolls smooth. 

flutter v1.9.1+hotfix.2, release mode, SingleChildScrollView has MultiTapGestureRecognizer can make this bug show

**Comment by BondarenkoStas** [comment]
@bigzhu Thank you for finding it! Please provide  a small self-contained app that demonstrates the problem. You can include small examples [inline if they're enclosed by backticks](https://help.github.com/en/articles/creating-and-highlighting-code-blocks), or just provide a public link to a github "gist" - https://gist.github.com/. In this way it will be possible to reproduce it exactly.

**Comment by bigzhu** [comment]
Created a simple project for the test again:  https://github.com/bigzhu/flutter_scroll_performance
you can checkout and run or copy code from https://gist.github.com/bigzhu/c40f86659b9fe97e0d77f36f8447a804

 I notice debug mode throw this error when scrolling:
```
════════ (140) Exception caught by scheduler library ═══════════════════════════════════════════════
'package:flutter/src/rendering/paragraph.dart': Failed assertion: line 859 pos 20: 'false': is not true.
════════════════════════════════════════════════════════════════════════════════════════════════════
```
Maybe this throw case release mode scrolls hard.

**Comment by chunhtai** [comment]
There are two issues here, the current implementation of textspan does not support MultiTapGestureRecognizer. It will throw assert error in debug mode and do not do any semantics update.

The other issue is textspan with too many gesture recognizer is slow when try to update semantics node. This apply to other GestureRecognizer as well.

The reason why it is ok in debug is because it crashes the semantics update, so no update is happening there.

The issue should already exist pre 1.7.8+hotfix.2. The only difference is we change the way it update when we have recognizer with no ontap method. If you use TapGestureRecognizer in  1.7.8+hotfix.2, you will see the same issue.

The code seems to try to assign a recognizer to every single word in the paragraph, which is very performance intensive. May I know what is the use case here?

**Comment by bigzhu** [comment]
@chunhtai  I built the app: [ebuoy](https://play.google.com/store/apps/details?id=net.bigzhu.english_buoy), it syncs YouTube Subtitle and combined to script article then learn the English word by word in this article.


<img src="https://user-images.githubusercontent.com/489815/65926478-fc581800-e427-11e9-97f3-04745b6b51b3.jpg" height=350/>

Every single word needs `onTap` to look up its meaning in the dictionary and `onLongTapDown` to tag this word is learned or unlearned, so I use `MultiTapGestureRecognizer`.

is there a better solution?

**Comment by chunhtai** [comment]
There are multiple ways can improve this, I am just listing a few..

1, If you don't care about accessibility, which you SHOULD. The easiest way is to wrap all widget with  [excludeSemantics](https://api.flutter.dev/flutter/widgets/ExcludeSemantics-class.html) However, this will make the entire paragraph unaccessible. It might hurt the user experience

2, Separating the paragraph into smaller sentence. The issue here is we need to update the semantics of the entire paragraph whenever the scroll offset changes. make it a smaller portion will help ease the load
an example can be this[this](https://gist.github.com/chunhtai/b59be1fce4200b2e7dd99f8000dd1958)

3, while 2 might improve the performance, but it might not be good enough. Ultimately we don't need ontap for every words if we can build a one ontap callback that will return which word in the paragraph are tapped. This will require you to build a custom widget for it. A possible implementation can be found in [here](https://gist.github.com/chunhtai/7041cb818162e244b64adccd61a5024f) This is just a draft, you will have to fill in the gap

**Comment by bigzhu** [comment]
using plan 2 when every sentence in itself `RichText`, the article scroll is smooth.

I truly appreciate your patience in helping me.

**Comment by chunhtai** [comment]
I am glad I was able to help. I will keep this ticket open to track the inconsistent performance between 1.7.8+hotfix.2 and v1.9.1+hotfix.2 when there is no ontap attachs

**Comment by chunhtai** [comment]
https://github.com/flutter/flutter/issues/41923 for multitap support

**Comment by liyuqian** [comment]
What's the current status of this regression and #41923 ? @chunhtai

**Comment by chunhtai** [comment]
It is still unresolved. The regression is only happening in a corner case.

**Comment by liyuqian** [comment]
Is there any plan to close this in 2020?

**Comment by chunhtai** [comment]
This is not currently on my list. I don't have a plan for this yet

**Comment by bigzhu** [comment]
#41923 I just  simply add `return` code before `assert(false);` :
```dart 
...
 else if (info.recognizer is MultiTapGestureRecognizer) {
            return;
          } else {
            assert(false);
          }
```
in  `packages/flutter/lib/src/rendering/paragraph.dart` will avoid crash and `MultiTapGestureRecognizer` is working fine.
This is not the final solution but it works for me. @liyuqian @chunhtai

**Comment by liyuqian** [comment]
Also CC @dnfield who last modified the related code mentioned in https://github.com/flutter/flutter/issues/40480#issuecomment-577489177

**Comment by dnfield** [comment]
@GaryQian and I discussed this here - https://github.com/flutter/flutter/pull/36243#discussion_r304050280

The basic structure of the current code isn't really sustainable for adding new types of gesture recognizers.

What we really need is a way for a recognizer to tell us how to update a semantics configuration object. Otherwise we end up switching over types and every time someone adds a new type it will break.

My change was actually just moving existing code - this behavior was impelmented before I touched it.

**Comment by dnfield** [comment]
@chunhtai 's suggestions in https://github.com/flutter/flutter/issues/40480#issuecomment-537248850 are very good.

We should try to make this case work better for a render paragraph, but I'm not sure it will ever work as well as a custom widget in this case. This is a challenging problem to solve in a generic way, and you can see that in most graphical frameworks when it comes to RichText controls.

**Comment by dnfield** [comment]
> #41923 I just simply add `return` code before `assert(false);` :
> 
> ```dart
> ...
>  else if (info.recognizer is MultiTapGestureRecognizer) {
>             return;
>           } else {
>             assert(false);
>           }
> ```
> 
> in `packages/flutter/lib/src/rendering/paragraph.dart` will avoid crash and `MultiTapGestureRecognizer` is working fine.
> This is not the final solution but it works for me. @liyuqian @chunhtai

This avoids the crash but it means that your widget won't work with accessibility services.

**Comment by liyuqian** [comment]
According to https://github.com/flutter/flutter/issues/40480#issuecomment-577511423, I wonder if we shall remove the "regression" label, and add this issue to some of our future plans since this is not easily fixable by reverting some commit. We currently have too many "regression" issues, and we need to focus on the real ones that can be fixed quickly. CC @Hixie @kf6gpe

**Comment by dnfield** [comment]
This is not a regression - this never worked, it just used to not print an assertion. Before the assertion was added, this was just a runtime crash with no assertion.

**Comment by dnfield** [comment]
To be honest, I think we should just close this as a dupe of https://github.com/flutter/flutter/issues/41923.

**Comment by dnfield** [comment]
I'm going to close this - we have no meaningful way of measuring the performance of this code since it just flat out doesn't work today, and can't possibly work until #41923.  I'm closing on the supposition that the performance issues noted have to do with the fact that the framework actually lets you know that this code is broken, whereas it may have previously silently let it through until eventually letting it crash for someone using a11y services.

**Comment by liyuqian** [comment]
Thank you Dan! Yes, let's redirect further discussions to https://github.com/flutter/flutter/issues/41923
